### PR TITLE
Add tests against live `quetz`

### PR DESCRIPTION
**Pull request by simonbohnen**
- Add tests against a locally running `quetz` instance

### CODE REVIEW

File: src/quetz_client/client.py [hunk 1]
```diff
@@ -91,7 +94,8 @@ def yield_channel_members(self, channel: str) -> Iterator[ChannelMember]:
         response = self.session.get(url=url)
         response.raise_for_status()
         for member_json in response.json():
-            yield ChannelMember(**member_json)
+            print(member_json)
```
Comment (janjagusch) [inline src/quetz_client/client.py]:
```suggestion
```

File: tests/conftest.py [hunk 2]
```diff
@@ -43,106 +139,235 @@ def mock_default_paginated_empty(requests_mock, test_url):
...
-                "packages_count": 11,
+                "name": channels[i].name,
+                "description": channels[i].description,
+                "private": channels[i].private,
+                "size_limit": channels[i].size_limit,
+                "ttl": channels[i].ttl,
+                "mirror_channel_url": channels[i].mirror_channel_url,
+                "mirror_mode": channels[i].mirror_mode,
+                "members_count": channels[i].members_count,
+                "packages_count": channels[i].packages_count,
```
Comment (janjagusch) [inline tests/conftest.py]: shouldn't there be a way to turn a dataclass into a dict?

File: tests/conftest.py [hunk 3]
```diff
@@ -43,106 +139,235 @@ def mock_default_paginated_empty(requests_mock, test_url):
...
+
+@pytest.fixture(scope="module")
+def live_dave(live_users):
+    return get_user_with_username(live_users, "dave")
+
+
+def get_user_with_username(users, username):
+    users = [u for u in users if u.username == username]
+    assert len(users) == 1
+    return users[0]
```
Comment (janjagusch) [inline tests/conftest.py]:
```suggestion
    return {u.username: u for u in users}[username]
```

File: tests/test_client.py [hunk 4]
```diff
@@ -132,71 +185,92 @@ def test_from_token():
...
     with temporary_package_file() as file:
-        quetz_client.post_file_to_channel(channel, file)
+        live_client.post_file_to_channel("a", file)
 
-    # the last request here is the download of the test package file, thus we need to access the second-to-last request
-    last_request = requests_mock.request_history[1]
-    assert last_request.method == "POST"
+    packages = live_client.yield_packages("a")
+
+    assert any(p.name == "xtensor" for p in packages)
```
Comment (janjagusch) [inline tests/test_client.py]: Should we separate `mock` and `live` tests into separate files?

File: tests/test_client.py [hunk 5]
```diff
@@ -132,71 +185,92 @@ def test_from_token():
...
+def test_live_post_file_to_channel(
+    live_client: QuetzClient,
+    live_post_channel_a,
+    requests_mock,
+):
     requests_mock.register_uri(
         "GET",
         "https://conda.anaconda.org/conda-forge/linux-64/xtensor-0.16.1-0.tar.bz2",
         real_http=True,
     )
```
Comment (janjagusch) [inline tests/test_client.py]: why is this necessary?

File: tests/test_client.py [hunk 6]
```diff
@@ -1,28 +1,30 @@
...
-    assert {ChannelMember(**ecm) for ecm in expected_channel_members} == channel_members
+    channel_members = set(client.yield_channel_members(channel=channel))
+    assert {
+        from_dict(ChannelMember, ecm) for ecm in expected_channel_a_members
+    } == channel_members
 
 
-def test_yield_users(quetz_client: QuetzClient, expected_users):
-    users = list(quetz_client.yield_users())
+def test_yield_users(client: QuetzClient, expected_users):
```
Comment (janjagusch) [inline tests/test_client.py]: are these live or mock tests?

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -27,6 +27,11 @@ jobs:
             pytest-md
             pytest-emoji
 
+      - name: Install quetz
+        run: |
+          micromamba install -c conda-forge quetz
+          micromamba install -c conda-forge sqlalchemy=1.4.46
```
Comment (janjagusch) [inline .github/workflows/ci.yml]: We can probably remove the pin now.

**Review by janjagusch — COMMENTED** [review]
nice!

File: tests/conftest.py (hunk 2 above)
Comment (simonbohnen) [reply tests/conftest.py]: Ah good point!

File: tests/test_client.py (hunk 4 above)
Comment (simonbohnen) [reply tests/test_client.py]: Yes 👍🏼

File: tests/test_client.py (hunk 5 above)
Comment (simonbohnen) [reply tests/test_client.py]: I'm not sure. I would think that by using `real_http=True` in `live_client` in `conftest.py`, this would be enough, but it's not for some reason. I added a comment.

File: tests/test_client.py (hunk 6 above)
Comment (simonbohnen) [reply tests/test_client.py]: The tests which don't have `live` or `mock` in their names run once for each.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (simonbohnen) [reply .github/workflows/ci.yml]: [Not quite yet](https://github.com/mamba-org/quetz-client/actions/runs/4184517170/jobs/7250270553), but soon it will work 😅

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -22,11 +22,15 @@ jobs:
...
         with:
-          cache-env: true
+          cache-env: false
           extra-specs: |
             pytest-md
             pytest-emoji
 
+      - name: Install quetz
+        run: |
+          micromamba install -c conda-forge quetz
```
Comment (janjagusch) [inline .github/workflows/ci.yml]:
```suggestion
          micromamba install -c conda-forge 'quetz>=0.6'
```

**Comment by janjagusch** [comment]
@simonbohnen, could you please resolve the merge conflicts? then i will give this a final review and hopefully merge it. :)

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -27,6 +27,10 @@ jobs:
             pytest-md
             pytest-emoji
 
+      - name: Install quetz
+        run: |
+          micromamba install -c conda-forge 'quetz>=0.6'
```
Comment (janjagusch) [inline .github/workflows/ci.yml]:
```suggestion
          micromamba install -c conda-forge 'quetz>=0.6.1'
```

File: environment.yml [hunk 10]
```diff
@@ -17,7 +17,9 @@ dependencies:
   - pytest-mock
   - requests-mock
   - httpx
+  # Runtime dependencies
   - fire
+  - dacite
   - tbump
```
Comment (janjagusch) [inline environment.yml]: Why is `tbump` a runtime dependency?

File: tests/test_both.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
+from dacite import from_dict
```
Comment (janjagusch) [inline tests/test_both.py:1]: Can you quickly elaborate how these tests run both in mocked and live mode?

**Review by janjagusch — COMMENTED** [review]
almost there!

File: tests/test_both.py (hunk 11 above)
Comment (simonbohnen) [reply tests/test_both.py:1]: This fixture is run twice, once emitting the mock client and once emitting the live client:

https://github.com/mamba-org/quetz-client/blob/1180ccfa001b0ed3e894f6c4216ab2fd31cee31e/tests/conftest.py#L125-L127

File: environment.yml (hunk 10 above)
Comment (simonbohnen) [reply environment.yml]: Good point, I synced the dependencies, also with `setup.cfg`.

**Review by janjagusch — APPROVED** [review]
almost there!
